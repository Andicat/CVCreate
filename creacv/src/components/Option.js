import React from 'react';
import PropTypes from 'prop-types';

import {connect} from 'react-redux';
import {createOption} from './utils';
import {cvElement_update} from '../redux/cvDataAC';

class Option extends React.PureComponent {

    static propTypes = {
        elementId: PropTypes.string.isRequired,
        optionName: PropTypes.string,
        optionValue: PropTypes.any,
    };

    onChangeValue = (value) => {
        this.props.dispatch(cvElement_update(this.props.elementId,this.props.optionName,value));
    }

    render () {
        let optionCode = createOption(this.props.optionName,this.props.optionValue,this.onChangeValue);
        return (
            <div className='options__elem'>
                {optionCode}
            </div>
        );
    }
}

const mapStateToProps = function (state) {
    return {
        elementId: state.cvData.activeElementId, 
    };
  };

export default connect(mapStateToProps)(Option);
