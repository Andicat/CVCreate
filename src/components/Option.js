import React from 'react';
import PropTypes from 'prop-types';
import {createOption, createTooltipText} from '../modules/utils';

//опция, настройка
class Option extends React.PureComponent {

    static propTypes = {
        blockId: PropTypes.number.isRequired,
        optionName: PropTypes.string,
        optionValue: PropTypes.any,
        cbOnChange: PropTypes.func.isRequired,
    };

    static defaultProps = {
        blockId: 0,
    };

    state = {
        tooltip: null,
    }

    //при изменении значения
    onChangeValue = (value) => {
        this.props.cbOnChange(this.props.blockId,this.props.optionName,value);
    }

    //показ подсказки при наведении мыши
    onMouseOver = (evt) => {
        if (evt.target.getAttribute('data-tooltip')) {
            this.setState({tooltip: createTooltipText(this.props.optionName)});
        }
    }

    //убираем подсказку
    onMouseOut = () => {
        this.setState({tooltip: null});
    }

    render () {
        let optionCode = createOption(this.props.optionName,this.props.optionValue,this.onChangeValue);
        if (!optionCode) {
            return null;
        }
        return (
            <div className='option' onMouseOver={this.onMouseOver} onMouseOut={this.onMouseOut} onClick={this.onMouseOut}>
                {optionCode}
                {this.state.tooltip &&
                    <span className='option__tooltip'>{this.state.tooltip}</span>
                }
            </div>

        );
    }
}

export default Option;
