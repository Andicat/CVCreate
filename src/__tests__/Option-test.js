"use strict";

import {renderComponentWithStore, storeEmpty} from './utils-test';

import Option from '../components/Option';

let component;

let props = {blockId:2,cbOnChange:()=> true};

describe('Option render', () => {

    it('Цвет', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'color',optionValue:'#666'});
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Число', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'fontsize',optionValue:'18'});
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Кнопка', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'copy',optionValue:null});
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Чекбокс', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'bold',optionValue:true});
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Диапазон', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'opacity',optionValue:0.5});
        expect(component.toJSON()).toMatchSnapshot();
        let option = component.root.find( el => el.props.className=='option');
        option.props.onClick();
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Ссылка', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'file',optionValue:null});
        expect(component.toJSON()).toMatchSnapshot();
        let option = component.root.find( el => el.props.className=='option');
        option.props.onClick();
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('Список', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'font',optionValue:null});
        expect(component.toJSON()).toMatchSnapshot();
    });
});

describe('Option action', () => {

    it('Подсказка', () => {
        component = renderComponentWithStore(Option,storeEmpty,{...props,optionName:'color',optionValue:'#666'});
        let option = component.root.find( el => el.props.className=='option');
        const evt = {'target':{getAttribute:()=>{return true},}};
        option.props.onMouseOver(evt);
        expect(component.toJSON()).toMatchSnapshot();
    });
});
