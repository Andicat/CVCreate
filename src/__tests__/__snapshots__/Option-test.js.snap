// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option action Подсказка 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__color"
    data-tooltip={true}
    onChange={[Function]}
    type="color"
    value="#666"
  />
  <span
    className="option__tooltip"
  >
    color of text
  </span>
</div>
`;

exports[`Option render Диапазон 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__button option__down option__button--opacity"
    data-tooltip={true}
    onClick={[Function]}
    type="button"
  />
  <div
    className="option__drop-down"
    onMouseLeave={[Function]}
  >
    <div
      className="option__drop-down-line"
    >
      <input
        className="option__range"
        max={1}
        min={0}
        onInput={[Function]}
        step={0.01}
        type="range"
        value={0.5}
      />
    </div>
  </div>
</div>
`;

exports[`Option render Диапазон 2`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__button option__down option__button--opacity"
    data-tooltip={true}
    onClick={[Function]}
    type="button"
  />
  <div
    className="option__drop-down"
    onMouseLeave={[Function]}
  >
    <div
      className="option__drop-down-line"
    >
      <input
        className="option__range"
        max={1}
        min={0}
        onInput={[Function]}
        step={0.01}
        type="range"
        value={0.5}
      />
    </div>
  </div>
</div>
`;

exports[`Option render Кнопка 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__button option__button--copy"
    data-tooltip={true}
    onClick={[Function]}
    type="button"
  />
</div>
`;

exports[`Option render Список 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <select
    className="option__select"
    data-tooltip={true}
    onChange={[Function]}
    value={null}
  >
    <option
      style={
        Object {
          "fontFamily": "PTSans",
        }
      }
    >
      PTSans
    </option>
    <option
      style={
        Object {
          "fontFamily": "Roboto",
        }
      }
    >
      Roboto
    </option>
    <option
      style={
        Object {
          "fontFamily": "Helvetica",
        }
      }
    >
      Helvetica
    </option>
    <option
      style={
        Object {
          "fontFamily": "Garamond",
        }
      }
    >
      Garamond
    </option>
  </select>
</div>
`;

exports[`Option render Ссылка 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    accept="image/*"
    className="option__file"
    id="file"
    name="file"
    onChange={[Function]}
    type="file"
  />
  <label
    className="option__label option__label--file"
    data-tooltip={true}
    htmlFor="file"
  >
    Load Image
  </label>
</div>
`;

exports[`Option render Ссылка 2`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    accept="image/*"
    className="option__file"
    id="file"
    name="file"
    onChange={[Function]}
    type="file"
  />
  <label
    className="option__label option__label--file"
    data-tooltip={true}
    htmlFor="file"
  >
    Load Image
  </label>
</div>
`;

exports[`Option render Цвет 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__color"
    data-tooltip={true}
    onChange={[Function]}
    type="color"
    value="#666"
  />
</div>
`;

exports[`Option render Чекбокс 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    checked={true}
    className="option__checkbox option__checkbox--bold"
    id="bold"
    onChange={[Function]}
    type="checkbox"
  />
  <label
    className="option__label"
    data-tooltip={true}
    htmlFor="bold"
  />
</div>
`;

exports[`Option render Число 1`] = `
<div
  className="option"
  onClick={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <input
    className="option__button option__button--left"
    data-tooltip={true}
    onClick={[Function]}
    type="button"
    value="–"
  />
  <input
    className="option__number"
    data-tooltip={true}
    max={60}
    min={6}
    readOnly={true}
    type="text"
    value="18"
  />
  <input
    className="option__button option__button--right"
    data-tooltip={true}
    onClick={[Function]}
    type="button"
    value="+"
  />
</div>
`;
